File: src/auth/auth.service.ts
import {
  ConflictException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { PrismaService } from '../config/prisma/prisma.service';
import {
  SignInDto,
  SignUpDto,
  SignInResponseDto,
  SignOutResponseDto,
  SignUpResponseDto,
  UserResponseDto,
} from '../schemas';

@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly jwtService: JwtService,
  ) {}

  async signUp(signUpDto: SignUpDto): Promise<SignUpResponseDto> {
    const { email, password } = signUpDto;

    // Check if user already exists
    const existingUser = await this.prisma.user.findUnique({
      where: { email },
    });

    if (existingUser) {
      throw new ConflictException('User already exists');
    }

    // Hash password
    const hashedPassword = await bcrypt.hash(password, 10);

    // Create user
    const user = await this.prisma.user.create({
      data: {
        email,
        password: hashedPassword,
      },
    });

    return {
      user: {
        id: user.id,
        name: user.name || '',
        email: user.email,
      },
    };
  }

  async signIn(signInDto: SignInDto): Promise<SignInResponseDto> {
    const { email, password } = signInDto;

    // Get user by email
    const user = await this.prisma.user.findUnique({
      where: { email },
    });

    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    // Check password
    const isPasswordValid = await bcrypt.compare(password, user.password);
    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid credentials');
    }

    // Generate tokens
    const payload = { sub: user.id, email: user.email };
    const token = this.jwtService.sign(payload);
    const refreshToken = this.jwtService.sign(payload, { expiresIn: '7d' });

    return {
      token,
      refreshToken,
      user: {
        id: user.id,
        name: user.name || '',
        email: user.email,
      },
    };
  }

  async signOut(): Promise<SignOutResponseDto> {
    return {
      message: 'Successfully signed out',
    };
  }
}

File: internal/service/auth_service.go
package service

import (
	"errors"

	"github.com/google/uuid"
	"github.com/lamkn06/user-app-golang.git/internal/repository"
	"github.com/lamkn06/user-app-golang.git/pkg/api/request"
	"github.com/lamkn06/user-app-golang.git/pkg/api/response"
	"golang.org/x/crypto/bcrypt"
)

type AuthService interface {
	SignUp(req request.SignUpRequest) (response.SignUpResponse, error)
	SignIn(req request.SignInRequest) (response.SignInResponse, error)
	SignOut(token string) (response.SignOutResponse, error)
}

type DefaultAuthService struct {
	userRepository repository.UserRepository
	jwtService     JWTService
}

func NewAuthService(userRepository repository.UserRepository, jwtService JWTService) AuthService {
	return &DefaultAuthService{
		userRepository: userRepository,
		jwtService:     jwtService,
	}
}

func (s *DefaultAuthService) SignUp(req request.SignUpRequest) (response.SignUpResponse, error) {
	// Check if user already exists
	_, err := s.userRepository.GetUserByEmail(req.Email)
	if err == nil {
		return response.SignUpResponse{}, errors.New("user already exists")
	}

	// Hash password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return response.SignUpResponse{}, err
	}

	// Create user
	user := repository.UserEntity{
		Id:       uuid.New(),
		Email:    req.Email,
		Password: string(hashedPassword),
	}

	createdUser, err := s.userRepository.InsertUser(user)
	if err != nil {
		return response.SignUpResponse{}, err
	}

	return response.SignUpResponse{
		ID:    createdUser.Id.String(),
		Email: createdUser.Email,
	}, nil
}

func (s *DefaultAuthService) SignIn(req request.SignInRequest) (response.SignInResponse, error) {
	// Get user by email
	user, err := s.userRepository.GetUserByEmail(req.Email)
	if err != nil {
		return response.SignInResponse{}, errors.New("invalid credentials")
	}

	// Check password
	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password))
	if err != nil {
		return response.SignInResponse{}, errors.New("invalid credentials")
	}

	// Generate tokens
	token, err := s.jwtService.GenerateToken(user.Id, user.Email)
	if err != nil {
		return response.SignInResponse{}, err
	}

	refreshToken, err := s.jwtService.GenerateRefreshToken(user.Id)
	if err != nil {
		return response.SignInResponse{}, err
	}

	return response.SignInResponse{
		Token:        token,
		RefreshToken: refreshToken,
		User: response.NewUserResponse{
			ID:    user.Id.String(),
			Email: user.Email,
		},
	}, nil
}

func (s *DefaultAuthService) SignOut(token string) (response.SignOutResponse, error) {
	// In a real application, you would blacklist the token
	// For now, we just return success
	return response.SignOutResponse{
		Message: "Successfully signed out",
	}, nil
}
